ocaml/runtime/hash.c

```diff
@@ -282,10 +283,15 @@ CAMLprim value caml_hash(value count, value limit, value seed, value obj)
         /* Mix in the tag and size, but do not count this towards [num] */
         h = caml_hash_mix_uint32(h, Cleanhd_hd(Hd_val(v)));
         /* Copy fields into queue, not exceeding the total size [sz] */
-        for (i = 0, len = Wosize_val(v); i < len; i++) {
+        for (i = 0, len = Scannable_wosize_val(v); i < len; i++) {
           if (wr >= sz) break;
           queue[wr++] = Field(v, i);
         }
+        /* Mix the flat suffix */
+        for (i = Scannable_wosize_val(v), len = Wosize_val(v); i < len; i++) {
+          h = caml_hash_mix_intnat(h, Field(v, i));
```

Comment:
(I don't think float values should be mixed
like that, `caml_hash_mix_double` is
performing some kind of normalization.)

Reply:
In this PR, we could call `caml_hash_mix_double`. But in future PRs, there will be flat elements that are, say, `int64#` (or two `float32#`s packed together, or 4 `float16#s`, or a `float32#` followed by 2 `float16#s`). I'm worried there's not going to be a way to normalize these correctly.

* If we just normalize -0 to 0, then I think we could do that fairly easily even with arbitrary mixing of small floats.
* But I'm worried about NaN normalization. (Is that even needed? Two nans are never equal, so it's vacuously true that `nan1 = nan2` implies `hash(nan1) = hash(nan2)`.

Reply:
Sure, I think it is just safe to ignore that part of the block.

Reply:
After some discussion:
  * We actually don't need to do any normalization. That's because mixed blocks can never "compare as equal" &mdash; polymorphic compare raises &mdash; so there's no risk of defying the hash invariant.
  * But this is all very tricky. I'm just going to return a constant for now on mixed blocks and revisit this later.

Reply:
I'd like to change this to raise before moving mixed blocks out of `layouts_alpha.` But I'd like to do this in a separate PR to more easily assess performance impact of making `caml_hash` to not be `[@@noalloc]`. @stedolan would you mind reviewing the remaining runtime changes and approving if you think they look good (modulo the hash matter)?